File: Project/GOAP/Planner/AStarPlanner.cpp
#include "AStarPlanner.h"
#include "PlannerPlan.h"
#include "PlannerNodeBase.h"
#include "PlannerGraphBase.h"

PlannerPlan AStarPlanner::FindPlan( PlannerNodeBase* startNode, PlannerNodeBase* endNode, PlannerGraphBase* graph )
{
	graph->Reset();
	m_OpenList.Clear();

	std::vector<PlannerNodeBase*> neighbors;

	// 1. Add startnode to openList
	m_OpenList.AddNode(startNode);

	// 2. Process openList until plan found, or openList empty.
	PlannerNodeBase* currentNode = nullptr;
	while(m_OpenList.HasNext()){
		currentNode = m_OpenList.GetNext();

		//currentNode->tempIsCurrent = true;

		//graph->DebugPrint();
		// Check if current node is goal
		if(graph->IsFinished(currentNode, endNode)){
			std::vector<PlannerNodeBase*> steps;
			PlannerNodeBase* parent = currentNode;
			while(parent){
				steps.push_back(parent);
				parent = parent->Parent;
			}

			//return PlannerPlan(std::move(std::vector<PlannerNodeBase*>()));
			//graph->DebugPrint();
			return graph->BuildPlan(steps);
		}

		// 3. Process each neighbor
		neighbors = graph->GetNeighbors(currentNode);
		for(PlannerNodeBase* neighbor : neighbors){
			//if(neighbor->State == PlannerNodeBase::CLOSED)
			//	continue;

			//neighbor->tempIsNeighbor = true;
			//graph->DebugPrint();
			//neighbor->tempIsNeighbor = false;

			// 4. Calculate cost so far (to compare if already in openList)
			float costSoFar = currentNode->CostSoFar + graph->CalculateCost(currentNode, neighbor);

			// 5. If this node is unvisited, or it's in the openList and is a better path, set data. Else skip.
			if(neighbor->State != PlannerNodeBase::NodeState::UNVISITED && costSoFar >= neighbor->CostSoFar)
				continue;
			else{
				neighbor->Parent = currentNode;
				neighbor->CostSoFar = costSoFar;
				neighbor->Heuristic = graph->CalculateHeuristic(neighbor, endNode);
				neighbor->EstimatedTotalCost = neighbor->CostSoFar + neighbor->Heuristic;
			}

			// 6. Add to openlist (if not already in list, in that case just re-sort CHANGED node only)
			if(neighbor->State == PlannerNodeBase::NodeState::UNVISITED)
				m_OpenList.AddNode(neighbor);
			else
				m_OpenList.Refresh(neighbor);
		}
		// tmp
		//currentNode->tempIsCurrent = false;
	}
	
	// No plan found, return empty
	return PlannerPlan(std::vector<PlannerNodeBase*>());
}

File: Project/GOAP/Planner/AStarPlanner.cs
using System.Collections.Generic;

namespace GOAP.Planner
{
	public class AStarPlanner<T> : IPlanner<T>
	{
		private IGraph<T> graph;
		private IPlannerList visitedNodes = new OpenList();

		public AStarPlanner(IGraph<T> graph)
		{
			this.graph = graph;
		}

		public T[] FindPlan(INode startNode, INode endNode)
		{
			var plan = new T[] { };
            visitedNodes.Clear();
			ResetNode(ref startNode);
			startNode.EstimatedTotalCost = float.MaxValue;

			visitedNodes.AddOrUpdate(startNode);
			while (visitedNodes.HasNext())
			{
				var currentNode = visitedNodes.PopNext();
				currentNode.IsClosed = true;

				if (graph.IsDone(currentNode, endNode))
				{
					plan = graph.BuildPlan(currentNode);
					break;
				}

				var neighbors = graph.GetNeighbors(currentNode);

				foreach (var neighbor in neighbors)
				{
					if (neighbor.IsClosed)
						continue;

					float costSoFar = currentNode.CostSoFar + currentNode.Cost;

					bool hasNotBeenVisited = !visitedNodes.HasNode(neighbor);
					bool isBetterPlan = costSoFar < neighbor.CostSoFar;
                    if (hasNotBeenVisited || isBetterPlan)
					{
						neighbor.Parent = currentNode;
						neighbor.CostSoFar = costSoFar;
						neighbor.EstimatedTotalCost = costSoFar + graph.CalculateEstimatedCost(neighbor, endNode);
						visitedNodes.AddOrUpdate(neighbor);
					}
				}
			}

			return plan;
		}

		private static void ResetNode(ref INode startNode)
		{
			startNode.Parent = null;
			startNode.CostSoFar = 0f;
			startNode.EstimatedTotalCost = 0f;
			startNode.IsClosed = false;
        }
	}
}